base_name_search_improved/models/ir_model.py

```diff
@@ -65,19 +156,111 @@ def name_search(self, name='', args=None,
                         domain = [(rec_name, operator, name.replace(' ', '%'))]
                         res = _extend_name_results(
                             self, base_domain + domain, res, limit)
+
                     # Try unordered word search on each of the search fields
-                    for rec_name in all_names:
-                        domain = [(rec_name, operator, x)
-                                  for x in name.split() if x]
+                    # we only perform this search if we have at least one
+                    # separator character
+                    separator = _get_separator(self)
+                    if separator in name:
+                        domain = []
+                        for word in name.split(separator):
+                            word_domain = []
+                            for rec_name in all_names:
+                                word_domain = (
```

Comment:
Thanks for the review and the suggestions @moylop260 !
Let me explain and please suggest me a better way to this if you have one. Let me show with an example (the same of the tests, it makes more sense with other case like the one explained on point 5 of this PR.
1. We have a partner named "Johann Gambolputty of Ulm" with phone "+351 777 333 555"
2. We want to search that partner and we remember his name start with "Johann" and his phone has "555"
3. Then we search "Johann 555"
4. I solve it by:
   1. splitting the search with the separator " "
   2. for each work we perform a search on each field with and or, for "Johann" it would be "['|', ('name', 'ilike', 'Johann'), ('ref', 'ilike', 'Johann')]
   3. we append each of those searches with and and, it would be: " ['&', '|', ('name', 'ilike', 'Johann'), ('ref', 'ilike', 'Johann'), '|', ('name', 'ilike', '555'), ('ref', 'ilike', '555')]

Any suggestion?